ownCloudAppShared/Intent/GetDirectoryListingIntentHandler.swift

```diff
@@ -0,0 +1,132 @@
+
+		if let path = intent.path {
+			completion(INStringResolutionResult.success(with: path))
+		} else {
+			completion(INStringResolutionResult.needsValue())
+		}
+
+	}
+
+	public func provideAccountOptions(for intent: GetDirectoryListingIntent, with completion: @escaping ([Account]?, Error?) -> Void) {
+		completion(OCBookmarkManager.shared.accountList, nil)
+	}
+
+	public func resolveAccount(for intent: GetDirectoryListingIntent, with completion: @escaping (AccountResolutionResult) -> Void) {
+		if let account = intent.account {
+			completion(AccountResolutionResult.success(with: account))
+		} else {
+			completion(AccountResolutionResult.needsValue())
+		}
+	}
+
+	public func resolveSortType(for intent: GetDirectoryListingIntent, with completion: @escaping (SortingTypeResolutionResult) -> Void) {
+		completion(SortingTypeResolutionResult.success(with: intent.sortType))
+	}
+
+	public func resolveSortDirection(for intent: GetDirectoryListingIntent, with completion: @escaping (SortingDirectionResolutionResult) -> Void) {
+		completion(SortingDirectionResolutionResult.success(with: intent.sortDirection))
+	}
+
+	public func handle(intent: GetDirectoryListingIntent, completion: @escaping (GetDirectoryListingIntentResponse) -> Void) {
+
+		if AppLockHelper().isPassCodeEnabled {
+			completion(GetDirectoryListingIntentResponse(code: .authenticationRequired, userActivity: nil))
+		} else {
+			if let path = intent.path, let uuid = intent.account?.uuid {
+				let accountBookmark = OCBookmarkManager.shared.bookmark(for: uuid)
+
+				if let bookmark = accountBookmark {
+					OCCoreManager.shared.requestCore(for: bookmark, setup: nil, completionHandler: { (core, error) in
+						if error == nil {
+							self.core = core
+							let targetDirectoryQuery = OCQuery(forPath: path)
+							targetDirectoryQuery.delegate = self
+
+							if targetDirectoryQuery.sortComparator == nil {
+								let sort = SortMethod(rawValue: (intent.sortType.rawValue - 1)) ?? SortMethod.alphabetically
+
+								targetDirectoryQuery.sortComparator = sort.comparator(direction: SortDirection(rawValue: (intent.sortDirection.rawValue - 1)) ?? SortDirection.ascendant)
+							}
+							core?.start(targetDirectoryQuery)
+						} else {
+							self.completion?(GetDirectoryListingIntentResponse(code: .failure, userActivity: nil))
+						}
+					})
+				} else {
+					completion(GetDirectoryListingIntentResponse(code: .accountFailure, userActivity: nil))
+				}
+			}
+
+			self.completion = completion
+		}
+	}
+
+	public func queryHasChangesAvailable(_ query: OCQuery) {
+		if query.state == .targetRemoved {
+			self.completion?(GetDirectoryListingIntentResponse(code: .pathFailure, userActivity: nil))
+			self.completion = nil
+		} else if query.state == .idle {
+			var directoryListing : [String] = []
+			if let results = query.queryResults {
+				for item in results {
```

Comment:
@hosy Could be simplified with `compactMap` / `append` of multiple elements at once

Reply:
Thanks!